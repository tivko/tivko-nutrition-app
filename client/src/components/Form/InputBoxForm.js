import {
  StyleSheet,
  Text,
  View,
  TextInput,
  TouchableOpacity,
  Alert,
} from "react-native";
import React, { useState } from "react";
import { Controller } from "react-hook-form";
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp,
} from "react-native-responsive-screen";
import { Feather } from "@expo/vector-icons";
import styles from "../../styles/login/InputBoxFormStyle";
import AlertComponent from "../Layout/AlertComponent";
import { bodyFatInfo } from "../../help/namesInUkrainian";
const InputBoxForm = ({
  control,
  name,
  rules = {},
  placeholder,
  label = true,
  secureTextEntry,
  keyboardType = "default",
}) => {
  const [isPasswordVisible, setIsPasswordVisible] = useState(true);
  const [modalVisible, setModalVisible] = useState(false);

  const togglePasswordVisibility = () => {
    setIsPasswordVisible(!isPasswordVisible);
  };
  return (
    <Controller
      control={control}
      name={name}
      rules={rules}
      render={({
        field: { value, onChange, onBlur },
        fieldState: { error },
      }) => (
        <View>
          {label && (
            <View style={styles.labelContainer}>
              <Text style={styles.labelText}>{placeholder}</Text>
            </View>
          )}
          <View
            style={[styles.container, { borderColor: error ? "red" : "grey" }]}
          >
            <TextInput
              value={
                value !== undefined && value !== null ? value.toString() : ""
              }
              onChangeText={(text) => onChange(text.trim())}
              onBlur={onBlur}
              placeholder={placeholder}
              style={styles.input}
              secureTextEntry={name === "password" && isPasswordVisible}
              keyboardType={keyboardType}
            />

            {name === "password" ? (
              <TouchableOpacity
                onPress={togglePasswordVisibility}
                style={styles.showPasswordContainer}
                testID="passwordToggle"
              >
                <Feather
                  name={isPasswordVisible ? "eye-off" : "eye"}
                  size={hp(2.5)}
                  color="black"
                />
              </TouchableOpacity>
            ) : (
              ""
            )}
            {name === "bodyfat" ? (
              <TouchableOpacity
                onPress={() => setModalVisible(!modalVisible)}
                style={styles.showPasswordContainer}
                testID="infoButton"
              >
                <Feather name={"info"} size={hp(2.5)} color="black" />
              </TouchableOpacity>
            ) : (
              ""
            )}
          </View>
          <View style={styles.errorContainer}>
            {error && (
              <Text style={styles.error}>{error.message || "Error"}</Text>
            )}
          </View>
          <AlertComponent
            modalVisible={modalVisible}
            setModalVisible={setModalVisible}
            info={bodyFatInfo}
            title={"Відсоток жиру"}
          />
        </View>
      )}
    />
  );
};

export default InputBoxForm;
