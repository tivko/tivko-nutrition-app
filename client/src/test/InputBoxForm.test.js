import React from "react";
import { TextInput, TouchableOpacity, Text } from "react-native";
import renderer from "react-test-renderer";
import InputBoxForm from "../components/Form/InputBoxForm";

jest.mock("../help/namesInUkrainian", () => ({
  bodyFatInfo: "Mocked Body Fat Info",
}));

jest.mock("react-hook-form", () => ({
  ...jest.requireActual("react-hook-form"),
  useForm: jest.fn(() => ({
    control: {
      _names: {},
    },
  })),
  Controller: jest.fn(({ render }) =>
    render({
      field: {
        value: "",
        onChange: jest.fn(),
        onBlur: jest.fn(),
      },
      fieldState: {},
    })
  ),
}));

describe("InputBoxForm", () => {
  it("Рендериться коректно зі стандартними параметрами", () => {
    const component = renderer.create(
      <InputBoxForm name="test" placeholder="Test Input" />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Перемикає видимість пароля", () => {
    const component = renderer.create(
      <InputBoxForm name="password" placeholder="Password" />
    );
    const instance = component.root;
    const input = instance.findByType(TextInput);
    const toggleButton = instance.findByProps({ testID: "passwordToggle" });
    toggleButton.props.onPress();
    expect(input.props.secureTextEntry).toBe(false);
    toggleButton.props.onPress();
    expect(input.props.secureTextEntry).toBe(true);
  });

  it("відкриває модальне вікно при натисканні на іконку інформації про відсоток жиру", () => {
    const component = renderer.create(
      <InputBoxForm name="bodyfat" placeholder="Body Fat" />
    );
    const instance = component.root;
    const infoButton = instance.findByProps({ testID: "infoButton" });
    infoButton.props.onPress();
  });
});
