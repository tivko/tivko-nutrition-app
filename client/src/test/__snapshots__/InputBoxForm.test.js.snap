// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputBoxForm renders correctly with default props 1`] = `
<View>
  <View
    style={
      {
        "paddingBottom": 5,
        "paddingTop": 8,
      }
    }
  >
    <Text
      style={
        {
          "color": "#808080",
          "fontSize": 26.5,
        }
      }
    >
      Test Input
    </Text>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "#ffff",
          "borderRadius": 5,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 80,
          "padding": 10,
          "width": 600,
        },
        {
          "borderColor": "grey",
        },
      ]
    }
  >
    <TextInput
      keyboardType="default"
      onBlur={[MockFunction]}
      onChangeText={[Function]}
      placeholder="Test Input"
      secureTextEntry={false}
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "fontSize": 24,
          "justifyContent": "center",
        }
      }
      value=""
    />
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "width": 600,
      }
    }
  />
</View>
`;

exports[`InputBoxForm Рендериться коректно зі стандартними параметрами 1`] = `
<View>
  <View
    style={
      {
        "paddingBottom": 5,
        "paddingTop": 8,
      }
    }
  >
    <Text
      style={
        {
          "color": "#808080",
          "fontSize": 26.5,
        }
      }
    >
      Test Input
    </Text>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "#ffff",
          "borderRadius": 5,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 80,
          "padding": 10,
          "width": 600,
        },
        {
          "borderColor": "grey",
        },
      ]
    }
  >
    <TextInput
      keyboardType="default"
      onBlur={[MockFunction]}
      onChangeText={[Function]}
      placeholder="Test Input"
      secureTextEntry={false}
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "fontSize": 24,
          "justifyContent": "center",
        }
      }
      value=""
    />
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "width": 600,
      }
    }
  />
</View>
`;
